### PR TITLE
4. simple prompt bot nodejs

### PR DESCRIPTION
**Pull request by benbrown**
For this sample, I started with the structure of the echobot sample, and mixed in @Stevenic's already comprehensive set of prompt samples from here:

https://github.com/Microsoft/botbuilder-js/blob/master/samples/dialogs/prompts-ts/src/app.ts

The end result is a bot that presents a menu of many prompt types, and allows you to loop through them and see them demonstrated.

- [ ] Needs deployment stuff
- [x] I need to check and see if there are other prompt types to demonstrate
- [x] Need to sync with @johnataylor for his .net version

### CODE REVIEW

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (vishwacsena) [inline javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: This sample needs to be simplified. Here's the [spec](https://github.com/Microsoft/BotBuilder-Samples/blob/v4/readme.md) - 
```markdown
Demonstrates prompt pattern by prompting user for a property. Introduces user state .vs. conversation state. Ask for name and prints back those information. 
```
So I'd recommend that we pare back this sample and simplify it to just do a name prompt.

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+} = require('botbuilder');
+
+const { 
+    DialogSet, TextPrompt, ConfirmPrompt, ChoicePrompt, DatetimePrompt, NumberPrompt, 
+    AttachmentPrompt
+} = require('botbuilder-dialogs');
+
+class MainDialog {
+    /**
+     *
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: extra line

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+                // if (choice.value === 'runAll') {
+                    // await dc.replace(choice.value);
+                // } else {
+                    await dc.context.sendActivity(`The demo will loop so say "menu" or "cancel" to end.`);
+                    await dc.replace('loop', { dialogId: choice.value });
+                // }
+            }
+        ]);
+
+
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: Extra blank lines.

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        this.dialogs.add('attachmentDemo', [
+            async function(dc) {
+                await dc.prompt('attachmentPrompt', `attachment: upload image(s)`);
+            },
+            async function(dc, values) {
+                await dc.context.sendActivity(MessageFactory.carousel(values, `Uploaded ${values.length} Attachment(s)`));
+                await dc.end();
+            }
+        ]);
+
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: Extra blank lines

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    }
+    /**
+     * 
+     * @param {Object} context on turn context object.
+     */
+    async onTurn(context) {
+        // see https://aka.ms/about-bot-activity-message to learn more about the message and other activity types
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: Should this have a "." at the end?

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+const restify = require('restify');
+const path = require('path');
+const ERROR = 1;
+
+// Import reuqired bot services. See https://ama.ms/bot-services to learn more about the different part of a bot
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: spelling. "required".

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: Spelling "ama.ms" != "aka.ms"

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: Grammer: "different part" -> "different parts" (plural).

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    console.log(`\n${server.name} listening to ${server.url}`);
+    console.log(`\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`);
+    console.log(`\nTo talk to your bot, open simple-prompt-bot.bot file in the Emulator`);
+});
+
+// .bot file path
+const BOT_FILE = path.join(__dirname, (process.env.botFilePath || ''));
+
+console.log('reading config from ', BOT_FILE);
+// read bot configuration from .bot file.
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: If you've got a "." at the end, then a capital letter should start the sentence. 

"Read".

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+// Note: Ensure you have a .env file and include botFilePath and botFileSecret.
+const ENV_FILE = path.join(__dirname, '.env');
+const env = require('dotenv').config({path: ENV_FILE});
+
+// Create HTTP server
+let server = restify.createServer();
+server.listen(process.env.port || process.env.PORT || 3978, function () {
+    console.log(`\n${server.name} listening to ${server.url}`);
+    console.log(`\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`);
+    console.log(`\nTo talk to your bot, open simple-prompt-bot.bot file in the Emulator`);
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: Emulator shouldn't be capitalized. 

Should have "." at the end of the sentences.

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+// Read botFilePath and botFileSecret from .env file
+// Note: Ensure you have a .env file and include botFilePath and botFileSecret.
+const ENV_FILE = path.join(__dirname, '.env');
+const env = require('dotenv').config({path: ENV_FILE});
+
+// Create HTTP server
+let server = restify.createServer();
+server.listen(process.env.port || process.env.PORT || 3978, function () {
+    console.log(`\n${server.name} listening to ${server.url}`);
+    console.log(`\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`);
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: grammer. Get the Bot Framework Emulator". (forgot "the")

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+// .bot file path
+const BOT_FILE = path.join(__dirname, (process.env.botFilePath || ''));
+
+console.log('reading config from ', BOT_FILE);
+// read bot configuration from .bot file. 
+let botConfig;
+try {
+    botConfig = BotConfiguration.loadSync(BOT_FILE, process.env.botFileSecret);
+} catch (err) {
+    console.log(`Error reading bot file. Please ensure you have valid botFilePath and botFileSecret set for your environment`);
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: "." at the end of the sentence.

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    console.log(err);
+    process.exit(ERROR);
+}
+
+// bot name as defined in .bot file 
+// See https://aka.ms/about-bot-file to learn more about .bot file its use and bot configuration .
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: grammer. Choose either:
[...] "more about .bot files" (plural) 
[...] "more about the .bot file".

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: extra space before the ".".

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+const BOT_CONFIGURATION = 'simple-prompt-bot';
+
+// Get bot endpoint configuration by service name
+const endpointConfig = botConfig.findServiceByNameOrId(BOT_CONFIGURATION);
+
+// Create adapter. See https://aka.ms/about-bot-adapter to learn more about .bot file its use and bot configuration .
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: same grammer comments

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+const adapter = new BotFrameworkAdapter({
+    appId: endpointConfig.appId || process.env.microsoftAppID,
+    appPassword: endpointConfig.appPassword || process.env.microsoftAppPassword
+});
+
+// Define state store for your bot. See https://aka.ms/about-bot-state to learn more about bots memory service
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: "." at the end. Misc grammer cleanup.

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+// A bot requires a sate store to priciest it dialog and user state between messages
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: spelling. "sate" != "state". 
Grammer. 
"priciest"? Not sure what's intended with that one.

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: priciest != persist.

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+const memoryStorage = new MemoryStorage();
+// CAUTION: The Memory Storage used here is for local bot debugging only. When the bot
+// is restarted, anything stored in memory will be gone. 
+// For production bots use the Azure CosmosDB storage, Azure Blob, or Azure Table storage provides.
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: "provides" != "providers".

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: We no longer support Azure Table Storage. Remove from comment.

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+});
+
+// Define state store for your bot. See https://aka.ms/about-bot-state to learn more about bots memory service
+// A bot requires a sate store to priciest it dialog and user state between messages
+const memoryStorage = new MemoryStorage();
+// CAUTION: The Memory Storage used here is for local bot debugging only. When the bot
+// is restarted, anything stored in memory will be gone. 
+// For production bots use the Azure CosmosDB storage, Azure Blob, or Azure Table storage provides. 
+// const { CosmosDbStorage } = require('botbuilder-azure');
+// const STORAGE_CONFIGURATION = 'cosmosDB'; // this is the name of the cosmos DB configuration in your .bot file
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: "cosmos DB" should be "CosmosDB".

File: javascript_nodejs/4.simple-prompt-bot/index.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+// const cosmosConfig = botConfig.findServiceByNameOrId(STORAGE_CONFIGURATION);
+// const cosmosStorage = new CosmosDbStorage({serviceEndpoint: cosmosConfig.connectionString, 
+//                                            authKey: ?, 
+//                                            databaseId: cosmosConfig.database, 
+//                                            collectionId: cosmosConfig.collection});
+
+// create conversation state with in-memory storage provider. 
+const conversationState = new ConversationState(memoryStorage);
+
+// register conversation state as a middleware. The ConversationState middleware automatically reads and writes conversation sate
```
Comment (cleemullins) [inline javascript_nodejs/4.simple-prompt-bot/index.js]: Register (capitalize) as you have a "." at the end of the sentence.

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: "sate" != "state".

Comment (cleemullins) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: "." at the end of the sentence.

**Comment by benbrown** [comment]
Hey guys, thanks for the feedback. I think this PR was a little premature as I hadn't finished the readme - apologies for that. On other projects, I have used open PRs to track work rather than signal completeness.

File: javascript_nodejs/4.simple-prompt-bot/.env [hunk 18]
```diff
@@ -0,0 +1,2 @@
+botFilePath=simple-prompt-bot.bot
```
[the remark this answers is not in the document]
Comment (benbrown) [reply javascript_nodejs/4.simple-prompt-bot/.env]: 👍

File: javascript_nodejs/4.simple-prompt-bot/README.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    ```bash
+    npm config set registry https://registry.npmjs.org/
+    ```
+
+# Testing the bot 
+
+Open the simple-prompt-bot.bot file in the Botframework Emulator.
+
+Once connected, the bot will present you with a menu of sample prompts. If it does not, send "menu"
+
```
[the remark this answers is not in the document]
Comment (benbrown) [reply javascript_nodejs/4.simple-prompt-bot/README.md]: 👍

File: javascript_nodejs/4.simple-prompt-bot/package.json [hunk 20]
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "console-echobot",
+  "version": "1.0.0",
+  "description": "Bot Builder v4 console echobot sample",
+  "main": "index.js",
+  "scripts": {
+    "start": "node ./index.js",
+    "test": "echo \"Error: no test specified\" && exit 1"
```
[the remark this answers is not in the document]
Comment (benbrown) [reply javascript_nodejs/4.simple-prompt-bot/package.json]: 👍

File: javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js (hunk 1 above)
Comment (benbrown) [reply javascript_nodejs/4.simple-prompt-bot/dialogs/mainDialog/index.js]: Doh! I hadn't seen this document.  Thanks.

File: javascript_nodejs/4.simple-prompt-bot/index.js (hunk 16 above)
Comment (stevengum) [reply javascript_nodejs/4.simple-prompt-bot/index.js]: Our references to "CosmosDB" are actually incorrect, we should be using "Cosmos DB" as indicated on [Azure](https://azure.microsoft.com/en-us/services/cosmos-db/).

**Comment by cleemullins** [comment]
@benbrown Is this PR ready to be merged in?

File: javascript_nodejs/4.simple-prompt-bot/package.json [hunk 21]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "console-echobot",
```
Comment (vishwacsena) [inline javascript_nodejs/4.simple-prompt-bot/package.json]: this neeeds to be simple-prompt-bot or something to that effect. Most certainly not console-echobot :)

File: javascript_nodejs/4.simple-prompt-bot/package.json [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+  "version": "1.0.0",
+  "description": "Bot Builder v4 console echobot sample",
```
Comment (vishwacsena) [inline javascript_nodejs/4.simple-prompt-bot/package.json]: same here. Need to change this to simple prompt bot.

**Comment by benbrown** [comment]
I implemented the suggested changes, and also added some more information to the readme to flesh out the description of the bot, and the scope of the available prompt classes.

**Pull request merged**